### PR TITLE
Firmware

### PR DESCRIPTION
**Pull request by JasonBrave**

### CODE REVIEW

File: actuator.c [hunk 1]
```diff
@@ -1,94 +1,24 @@
...
-    } else {
-        // Should not get here.
-        while (1) {}; // panic, this will be caught early on in testing
-    }
-    
-#ifdef INJECTOR
-    LATB4 = 1;
-    TRISB4 = 0; // set LIMIT_OPEN as output to power hall sensor
+uint8_t actuator_states = 0;
+uint8_t actuator_invert = 0;
```
Comment (BluCodeGH) [inline actuator.c]: You should get rid of all the actuator inverting logic. Inverting only makes sense if your CAN commands are OPEN/CLOSE, but our CAN commands are ON/OFF which is well-defined and should never be inverted. This caused a ton of confusion last year, lets pls avoid it this year.

File: error_checks.c [hunk 2]
```diff
@@ -5,35 +5,38 @@
 #include "mcc_generated_files/system/system.h"
 
 #include "error_checks.h"
-#include "board.h"
+// #include "board.h"
 #include "actuator.h"
 
+const float CONVERT_FACTOR = 1e+05 / 4096.0f * 3.3; // 100 V/V multiplier * mV conversion * (raw + continuity correction) / 12bit adc * vref
```
Comment (BluCodeGH) [inline error_checks.c]: This isn't correct. An adc raw output of 1 = 1/4096 * 3.3V into adc = 1/4096 * 3.3V / 100 across resistor, this is the inverse of that (and idk where the 1e5 came from). Also remember when you divide by a resistance in milliohms, you end up with mV / mohm = A, not mA! So you want to multiply the conversion factor by 1000 so you have microvolts.

File: error_checks.c [hunk 3]
```diff
@@ -5,35 +5,38 @@
...
 
 static bool battery_voltage_critical = false;
 
-bool check_battery_voltage_error(void){    //returns mV
-    adc_result_t batt_raw = ADCC_GetSingleConversion(channel_VBAT);
+bool check_battery_voltage_error(adcc_channel_t battery_channel) { // returns mV
+    adc_result_t batt_raw = ADCC_GetSingleConversion(battery_channel);
+    //adc_result_t batt_raw = 0;
 
     // Vref: 4.096V, Resolution: 12 bits -> raw ADC value is precisely in mV
```
Comment (BluCodeGH) [inline error_checks.c]: This comment (and thus your conversion math) is incorrect.

File: main.c [hunk 4]
```diff
@@ -1,34 +1,94 @@
...
 #include "error_checks.h"
 #include "i2c.h"
 #include "sensor_general.h"
-#include "actuator.h"
-#include "board.h"
 
 #include <xc.h>
 
-#define MAX_BUS_DEAD_TIME_ms 1000
+#define MAX_BUS_DEAD_TIME_ms 1000 // Unused
```
Comment (BluCodeGH) [inline main.c]: This isn't unused?

File: main.c [hunk 5]
```diff
@@ -74,137 +129,231 @@ int main(int argc, char **argv) {
...
+    actuator_init(1);
+#endif
 
-    uint32_t last_message_millis = 0; // last time we saw a can message
+    /*
+      Vent and Fill Dump: 0 to open, 1 to close (invert)
+      Injector: 0 to close, 1 to open
+    */
+
+    uint32_t last_message_millis = 0; // last time we saw a can message // FIXME reset
```
Comment (BluCodeGH) [inline main.c]: FIXME?

File: main.c [hunk 6]
```diff
@@ -74,137 +129,231 @@ int main(int argc, char **argv) {
...
-            status_ok &= check_actuator_pin_error(requested_actuator_state);
+            status_ok &= check_battery_voltage_error(batt_vol_sense);
+
+            status_ok &= check_5v_current_error(current_sense_5v); 
+            status_ok &= check_12v_current_error(current_sense_12v);
 
             // if there was an issue, a message would already have been sent out
-            if (status_ok) { send_status_ok(); }
+            if (status_ok) {
+                send_status_ok();
```
Comment (BluCodeGH) [inline main.c:191]: You've got two blocks of duplicated code checking for errors and sending status ok, only have one pls :smiling_face_with_three_hearts:

File: main.c [hunk 7]
```diff
@@ -74,137 +129,231 @@ int main(int argc, char **argv) {
...
             // Set safe state if:
             // 1. We haven't heard CAN traffic in a while
             // 2. We're low on battery voltage
             // "thread safe" because main loop should never write to requested_actuator_state
-            if (SAFE_STATE_ENABLED && (
-                    (millis() - last_command_millis > MAX_CAN_IDLE_TIME_MS)
-                    || is_batt_voltage_critical())) {
-                actuator_send_status(SAFE_STATE);
-                actuator_set(SAFE_STATE);
+#ifdef JASON
```
Comment (BluCodeGH) [inline main.c]: Remove this

File: main.c [hunk 8]
```diff
@@ -74,137 +129,231 @@ int main(int argc, char **argv) {
...
 
-            // With low-pass filter, uncomment if low-pass filtering required:
-            uint16_t pressure_4_20_psi = update_pressure_psi_low_pass();
-
-            can_msg_t sensor_msg;
-            build_analog_data_msg(millis(), PT_SENSOR_ID, pressure_4_20_psi, &sensor_msg);
-            txb_enqueue(&sensor_msg);
-        }
-#endif
+// FIXME wtf is this. PRES_TIME_DIFF_ms is defined in the header file to be 250
```
Comment (BluCodeGH) [inline main.c]: Remove?

File: main.c [hunk 9]
```diff
@@ -74,137 +129,231 @@ int main(int argc, char **argv) {
...
+            last_pres_ox_millis = millis();
+            uint16_t pressure_ox_psi = update_pressure_psi_low_pass(pres_ox);
+            can_msg_t sensor_msg;
+            build_analog_data_msg(millis(), SENSOR_PRESSURE_OX, pressure_ox_psi, &sensor_msg);
+            txb_enqueue(&sensor_msg);
+        }
+#endif
         // send any queued CAN messages
         txb_heartbeat();
     }
```
Comment (BluCodeGH) [inline main.c:357]: Github wont let me comment lower but remove the random send_status_ok thats after the while loop, we'll never get there.

File: main.c [hunk 10]
```diff
@@ -238,22 +387,38 @@ static void can_msg_handler(const can_msg_t *msg) {
         return;
     }
 
+    // make able to handle multiple actuator
     switch (msg_type) {
         case MSG_GENERAL_CMD:
             cmd_type = get_general_cmd_type(msg);
             if (cmd_type == BUS_DOWN_WARNING) {
```
Comment (BluCodeGH) [inline main.c]: Not sure why this code ever existed but we should only ever be changing actuators in response to error conditions or actuator commands, not a bus down warning. Just ignore this message.

**Review by BluCodeGH — CHANGES_REQUESTED** [review]

File: defmplabxtrace.log
Comment (KavinSatheeskumar) [inline defmplabxtrace.log:1]: Don't include log files in your PR

File: actuator.h [hunk 11]
```diff
@@ -1,16 +1,12 @@
...
+#include <stdbool.h>
 
 void actuator_init();
-
-void actuator_set(enum ACTUATOR_STATE state);
-
-enum ACTUATOR_STATE get_actuator_state(void);
-
-void actuator_send_status(enum ACTUATOR_STATE req_state);
+void actuator_set(enum ACTUATOR_STATE state, uint8_t pin_num);
```
Comment (KavinSatheeskumar) [inline actuator.h:8]: This is nitpicky, but call it set_actuator_state, just so the naming convention lines up

File: actuator.c [hunk 12]
```diff
@@ -1,56 +1,50 @@
...
+    uint8_t state = (actuator_states >> pin_num) & 0x01;
+    
+    if (state == 1)
+        return ACTUATOR_ON;
+    
+    return ACTUATOR_OFF;
+    
+    
+    
+    /**
```
Comment (KavinSatheeskumar) [inline actuator.c]: Removed commented out code

File: actuator.c [hunk 13]
```diff
@@ -1,56 +1,50 @@
...
+        (state == ACTUATOR_ON) ? LED_ON_R() : LED_OFF_R(); 
+    }
+    else { // Vent/Fill
+        (state == ACTUATOR_ON) ? LED_ON_B() : LED_OFF_B(); 
     }
 }
 
-enum ACTUATOR_STATE get_actuator_state(void) {
+// TODO update this
+enum ACTUATOR_STATE get_actuator_state(uint8_t pin_num) {
```
Comment (KavinSatheeskumar) [inline actuator.c:34]: Function note used

File: IOExpanderDriver.c [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+#include <stdlib.h>
+
+#define PCA_ADDRESS 0x41
+
+#define INPUT 0x00
+#define OUTPUT 0x01
+#define POLARITY_REG 0x02
+#define CONFIGURATION 0x03
+
+void pca_init() {
```
Comment (KavinSatheeskumar) [inline IOExpanderDriver.c]: Function never used also is empty

File: actuator.c [hunk 15]
```diff
@@ -1,56 +1,50 @@
...
-#ifdef INJECTOR
-    LATB4 = 1;
-    TRISB4 = 0; // set LIMIT_OPEN as output to power hall sensor
-
-    TRISB3 = 1; // set LIMIT_CLOSED (pin 24) as input
-    ANSELB3 = 1; // set as analog input
-#else
-    TRISB4 = 1; // set LIMIT_OPEN (pin 25) as input
-    ANSELB4 = 0; // set as digital input
+uint8_t actuator_states = 0;
```
Comment (KavinSatheeskumar) [inline actuator.c:10]: We never actually read actuator states, so we should probs remove this variable and just have the set use the pca, no need to maintain the state of the actuator in the code

File: actuator.c [hunk 16]
```diff
@@ -1,56 +1,50 @@
...
 
-    TRISB3 = 1; // set LIMIT_CLOSED (pin 24) as input
-    ANSELB3 = 0; // set as digital input
-#endif
+void actuator_init() {
```
Comment (KavinSatheeskumar) [inline actuator.c:12]: If we are not using actuator_state, we can get rid of this variable.

Also, even if we do, we can probs add some logic to get_actator_state which avoids an init

**Review by KavinSatheeskumar — CHANGES_REQUESTED** [review]

File: actuator.c [hunk 17]
```diff
@@ -1,94 +1,44 @@
...
-    txb_enqueue(&hall_msg);
-#endif
-    enum ACTUATOR_STATE curr_state = get_actuator_state();
+// TODO update this
+enum ACTUATOR_STATE get_actuator_state(uint8_t pin_num) {
+
+    // temp code
+    uint8_t state = (actuator_states >> pin_num) & 0x01;
+
+    if (state == 1) {
```
Comment (KavinSatheeskumar) [inline actuator.c:39]: we could use ternary operator instead of this iff condition

`return (state == 1) ? ACTUATOR_ON : ACTUATOR_OFF`

**Review by KavinSatheeskumar — APPROVED** [review]

**Pull request merged**